src/libsyntax/visit_mut.rs

```diff
@@ -0,0 +1,886 @@
+    fn visit_enum_def(&mut self, enum_definition: &mut EnumDef,
+                      generics: &mut Generics, item_id: NodeId, _: Span) {
+        walk_enum_def(self, enum_definition, generics, item_id)
+    }
+    fn visit_variant(&mut self, v: &mut Variant, g: &mut Generics, item_id: NodeId) {
+        walk_variant(self, v, g, item_id)
+    }
+    fn visit_label(&mut self, label: &mut Label) {
+        walk_label(self, label)
+    }
+    fn visit_lifetime(&mut self, lifetime: &mut Lifetime) {
+        walk_lifetime(self, lifetime)
+    }
+    fn visit_mac(&mut self, _mac: &mut Mac) {
+        panic!("visit_mac disabled by default");
+        // N.B., see note about macros above.
+        // if you really want a visitor that
+        // works on macros, use this
+        // definition in your trait impl:
+        // visit::walk_mac(self, _mac)
+    }
+    fn visit_mac_def(&mut self, _mac: &mut MacroDef, _id: NodeId) {
+        // Nothing to do
+    }
+    fn visit_path(&mut self, path: &mut Path, _id: NodeId) {
+        walk_path(self, path)
+    }
+    fn visit_use_tree(&mut self, use_tree: &mut UseTree, id: NodeId, _nested: bool) {
+        walk_use_tree(self, use_tree, id)
+    }
+    fn visit_path_segment(&mut self, path_span: Span, path_segment: &mut PathSegment) {
+        walk_path_segment(self, path_span, path_segment)
+    }
+    fn visit_generic_args(&mut self, path_span: Span, generic_args: &mut GenericArgs) {
+        walk_generic_args(self, path_span, generic_args)
+    }
+    fn visit_generic_arg(&mut self, generic_arg: &mut GenericArg) {
+        match generic_arg {
+            GenericArg::Lifetime(lt) => self.visit_lifetime(lt),
+            GenericArg::Type(ty) => self.visit_ty(ty),
+        }
+    }
+    fn visit_assoc_type_binding(&mut self, type_binding: &mut TypeBinding) {
+        walk_assoc_type_binding(self, type_binding)
+    }
+    fn visit_attribute(&mut self, attr: &mut Attribute) {
+        walk_attribute(self, attr)
+    }
+    fn visit_tt(&mut self, tt: TokenTree) {
+        walk_tt(self, tt)
+    }
+    fn visit_tts(&mut self, tts: TokenStream) {
+        walk_tts(self, tts)
+    }
+    fn visit_token(&mut self, _t: Token) {}
+    // FIXME: add `visit_interpolated` and `walk_interpolated`
+    fn visit_vis(&mut self, vis: &mut Visibility) {
+        walk_vis(self, vis)
+    }
+    fn visit_fn_ret_ty(&mut self, ret_ty: &mut FunctionRetTy) {
+        walk_fn_ret_ty(self, ret_ty)
+    }
+}
+
+macro_rules! walk_list_mut {
+    ($visitor: expr, $method: ident, $list: expr $(, $extra_args: expr),*) => {
+        let mut i = 0;
+        loop {
+            if i == $list.len() {
+                break;
+            }
+
+            if let Some(replacement) = $visitor.$method(&mut $list[i], $(, $extra_args)*) {
+                $list.remove(i);
+                let rlen = replacement.len();
+                for (j, r) in replacement.into_iter().enumerate() {
+                    $list.insert(i + j, r);
```

Comment:
Also, don't we have a range insert, like `list.insert(replacement.into_iter())` or something?